knnaas/worker/KNNIndex.cpp

```diff
@@ -0,0 +1,415 @@
+		    }
+		    
+            return res;
+	}
+
+    void calculate_distance_return_list_using_iterator(
+                                                        int thread_id, 
+                                                        std::vector<float>& input_vector,
+                                                        std::vector<long>& input_tags,
+                                                        long begin_index,
+                                                        long end_index,
+                                                        std::vector<ResultRecord>& result,
+                                                        int top_n
+                                                       ) {
+            float dist;
+            long counter = 0;
+		    IndexType::iterator index_iterator = index.begin();
+            std::vector<float>::iterator index_feature_iterator, input_vector_iterator;
+                //for (; (indexIterator != index.end()) && counter++ < (endIndex - beginIndex); indexIterator++) { 
+		    int num_loops = end_index - begin_index;
+            std::cout << "[DIST CAL STARTED]" << std::endl;  
+            for (std::advance(index_iterator, begin_index); counter++ < num_loops; index_iterator++) { 
+			    bool tags_matched = true;
+                
+			    if (index_iterator->deleted == true) {
+                    std::cout << "[DELETED]" << std::endl;
+				    continue;
+			    }
+
+                if (input_vector.size() != index_iterator->vector.size()) {
+                    std::cout << "[Dim Mismatch]" << std::endl ;
+                    std::cout << std::to_string(input_vector.size()); 
+                    std::cout << "[Input Vector]" << std::endl;
+                    std::cout << std::to_string(index_iterator->vector.size());
+                    std::cout << "[Iterator Vector]" << std::endl;
+                    continue;
+                }
+			
+                for (auto it = input_tags.begin(); it != input_tags.end(); it++) {
+				
+                    if (index_iterator->tags.find(*it) == index_iterator->tags.end()) {
+                        std::cout << "[TAG MISMATCHED]" << std::endl;
+					    tags_matched = false;
+					    break;
+				    }
+			    }
+			
+                if (tags_matched == false) {
+				    continue;
+			    }   
+            
+                for (
+                    dist = 0, 
+                    index_feature_iterator = index_iterator->vector.begin(), 
+                    input_vector_iterator = input_vector.begin(); 
+                  
+                    input_vector_iterator != input_vector.end();
+
+                    index_feature_iterator++, 
+                    input_vector_iterator++
+                    ) {
+                
+                    dist +=  (*index_feature_iterator - *input_vector_iterator) * (*index_feature_iterator - *input_vector_iterator);
+                }
+            
+                float x = sqrt(dist);
```

Comment:
Let it be for now.
1. We return the euclidean distance. That is the contract.
2. More importantly, make the change only if we can benchmark and demonstrate significant improvement.